File: src/lib/annis/json/jsonqueryparser.cpp
/*
   Copyright 2017 Thomas Krause <thomaskrause@posteo.de>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

#include "jsonqueryparser.h"
#include <annis/annosearch/exactannokeysearch.h>    // for ExactAnnoKeySearch
#include <annis/annosearch/exactannovaluesearch.h>  // for ExactAnnoValueSearch
#include <annis/annosearch/regexannosearch.h>       // for RegexAnnoSearch
#include <annis/operators/dominance.h>              // for Dominance
#include <annis/operators/identicalcoverage.h>      // for IdenticalCoverage
#include <annis/operators/identicalnode.h>
#include <annis/operators/inclusion.h>              // for Inclusion
#include <annis/operators/overlap.h>                // for Overlap
#include <annis/operators/pointing.h>               // for Pointing
#include <annis/operators/precedence.h>             // for Precedence
#include <annis/operators/partofsubcorpus.h>
#include <assert.h>                                 // for assert
#include <re2/re2.h>                                // for RE2
#include <limits>                                   // for numeric_limits
#include <map>                                      // for _Rb_tree_const_it...
#include <utility>                                  // for pair
#include "annis/db.h"                               // for DB
#include "annis/json/json.h"                        // for Value, ValueConst...
#include "annis/query/query.h"                            // for Query
#include "annis/queryconfig.h"                      // for QueryConfig
#include "annis/stringstorage.h"                    // for StringStorage
#include "annis/types.h"                          // for Edge, GraphStatistic

#include <boost/optional.hpp>

#include <functional>

using namespace annis;

JSONQueryParser::JSONQueryParser()
{
}
std::shared_ptr<Query> JSONQueryParser::parse(DB &db, std::istream& jsonStream, const QueryConfig config)
{
  return parse(db, db.f_getGraphStorage, db.f_getAllGraphStorages, jsonStream, config);
}

std::shared_ptr<Query> JSONQueryParser::parse(const DB& db, DB::GetGSFuncT getGraphStorageFunc,
                                              DB::GetAllGSFuncT getAllGraphStorageFunc,
                                              std::istream& jsonStream, const QueryConfig config)
{
  std::vector<std::shared_ptr<SingleAlternativeQuery>> result;

  // parse root as value
  Json::Value root;
  jsonStream >> root;

  // iterate over all alternatives
  const auto& alternatives = root["alternatives"];
  for(const auto& alt : alternatives)
  {
    std::shared_ptr<SingleAlternativeQuery> q = std::make_shared<SingleAlternativeQuery>(db, config);


    // add all nodes
    const auto& nodes = alt["nodes"];

    std::map<std::uint64_t, size_t> nodeIdToPos;
    boost::optional<size_t> firstNodePos;
    for (auto it = nodes.begin(); it != nodes.end(); it++)
    {
      auto& n = *it;
      size_t pos = parseNode(db, n, getGraphStorageFunc, q);
      nodeIdToPos[std::stoull(it.name())] = pos;
      if(!firstNodePos)
      {
        firstNodePos = pos;
      }
    }

    // add all joins
    const auto& joins = alt["joins"];
    for (auto it = joins.begin(); it != joins.end(); it++)
    {
      parseJoin(db, getGraphStorageFunc, getAllGraphStorageFunc, *it, q, nodeIdToPos);
    }

    // add all meta-data
    const auto& meta = alt["meta"];
    boost::optional<size_t> firstMetaIdx = boost::none;
    for (auto it = meta.begin(); it != meta.end(); it++)
    {
      auto& m = *it;

      // add an artificial node that describes the document/corpus node
      size_t metaNodeIdx = addNodeAnnotation(db, q, optStr(m["namespace"]),
            optStr(m["name"]), optStr(m["value"]),
            optStr(m["textMatching"]), true);

      if(firstMetaIdx)
      {
        // avoid nested loops by joining additional meta nodes with a "identical node"
        q->addOperator(std::make_shared<IdenticalNode>(db), metaNodeIdx, *firstMetaIdx);

      }
      else
      {
        firstMetaIdx = metaNodeIdx;
        // add a special join to the first node of the query
        q->addOperator(std::make_shared<PartOfSubCorpus>(getGraphStorageFunc, db.strings, 1),
          metaNodeIdx, *firstNodePos);

      }
    }

    result.push_back(q);

  } // end for each alternative
  return std::make_shared<Query>(result);
}

std::shared_ptr<Query> JSONQueryParser::parseWithUpgradeableLock(DB &db,
                                                                 std::string queryAsJSON, boost::upgrade_lock<DBLoader>& lock,
                                                                 const QueryConfig config)
{
  DB::GetAllGSFuncT allFunc = [&db,&lock](ComponentType type, const std::string &name)
  {
    if(db.allGraphStoragesLoaded(type, name))
    {
      return db.getAllGraphStorages(type, name);
    }
    else
    {
      // loading the graph storages needs a unique lock to ensure nobody else is loading the same GS at the same time
      boost::upgrade_to_unique_lock<DBLoader> uniqueLock(lock);
      return db.getAllGraphStorages(type, name);
    }
  };

  DB::GetGSFuncT func = [&db,&lock](ComponentType type, const std::string &layer, const std::string &name)
  {
    if(db.isGraphStorageLoaded(type, layer, name))
    {
      return db.getGraphStorage(type, layer, name);
    }
    else
    {
      // loading the graph storages needs a unique lock to ensure nobody else is loading the same GS at the same time
      boost::upgrade_to_unique_lock<DBLoader> uniqueLock(lock);
      return db.getGraphStorage(type, layer, name);
    }
  };

  std::stringstream ss;
  ss << queryAsJSON;
  return annis::JSONQueryParser::parse(db, func, allFunc, ss, config);
}

size_t JSONQueryParser::parseNode(const DB& db, const Json::Value node, DB::GetGSFuncT getGraphStorageFunc, std::shared_ptr<SingleAlternativeQuery> q)
{

  // annotation search?
  if (node["nodeAnnotations"].isArray() && node["nodeAnnotations"].size() > 0)
  {
    // get the first one
    auto nodeAnno = node["nodeAnnotations"][0];

    return addNodeAnnotation(db, q, optStr(nodeAnno["namespace"]),
      optStr(nodeAnno["name"]), optStr(nodeAnno["value"]),
      optStr(nodeAnno["textMatching"]));

  }// end if annotation search
  else
  {
    // check for special non-annotation search constructs
    // token search?
    if (node["spannedText"].isString()
        || (node["token"].isBool() && node["token"].asBool()))
    {
      size_t n_pos = addNodeAnnotation(db, q, optStr(annis_ns), optStr(annis_tok),
                                       optStr(node["spannedText"]),
                                       optStr(node["spanTextMatching"]), true);

      // special treatment for explicit searches for token (tok="...)
      if(node["token"].isBool() && node["token"].asBool())
      {
        std::shared_ptr<const ReadableGraphStorage> covGS = getGraphStorageFunc(ComponentType::COVERAGE, annis_ns, "");
        if(covGS)
        {
          q->addFilter(n_pos, [&db, covGS] (const Match& m) -> bool
          {
            return static_cast<bool>(db.nodeAnnos.getAnnotations(m.node, db.getNamespaceStringID(), db.getTokStringID()))
                && covGS->getOutgoingEdges(m.node).empty();
          }, "is_token");
        }
      }

      return n_pos;
    }// end if token has spanned text
    else
    {
      // just search for any node
      return addNodeAnnotation(db, q, optStr(annis_ns), optStr(annis_node_type),
        boost::optional<std::string>("node"), boost::optional<std::string>("EXACT_EQUAL"));
    }
  } // end if special case

}

size_t JSONQueryParser::addNodeAnnotation(const DB& db,
  std::shared_ptr<SingleAlternativeQuery> q,
  boost::optional<std::string> ns,
  boost::optional<std::string> name,
  boost::optional<std::string> value,
  boost::optional<std::string> textMatching,
  bool wrapEmptyAnno)
{

  if (value)
  {
    bool exact = *textMatching == "EXACT_EQUAL";
    bool regex = *textMatching == "REGEXP_EQUAL";
    if(regex)
    {
      if(canReplaceRegex(*value))
      {
        exact = true;
      }
    }
    
    // search for the value
    if (exact)
    {
      // has namespace?
      if (ns)
      {
        return q->addNode(std::make_shared<ExactAnnoValueSearch>(db,
          *ns,
          *name,
          *value),
          wrapEmptyAnno);
      }
      else
      {
        return q->addNode(std::make_shared<ExactAnnoValueSearch>(db,
          *name,
          *value),
          wrapEmptyAnno);
      }
    }
    else if (regex)
    {
      // has namespace?
      if (ns)
      {
        return q->addNode(std::make_shared<RegexAnnoSearch>(db,
          *ns,
          *name,
          *value),
          wrapEmptyAnno);
      }
      else
      {
        return q->addNode(std::make_shared<RegexAnnoSearch>(db,
          *name,
          *value),
          wrapEmptyAnno);
      }
    }

  }// end if has value
  else
  {
    // only search for key
    // has namespace?
    if (ns)
    {
      return q->addNode(std::make_shared<ExactAnnoKeySearch>(db,
        *ns,
        *name),
        wrapEmptyAnno);
    }
    else
    {
      return q->addNode(std::make_shared<ExactAnnoKeySearch>(db,
        *name),
        wrapEmptyAnno);
    }
  }
  // should never happen
  assert(false);
  return 0;
}

void JSONQueryParser::parseJoin(const DB& db,
                                DB::GetGSFuncT getGraphStorageFunc,
                                DB::GetAllGSFuncT getAllGraphStorageFunc,
                                const Json::Value join, std::shared_ptr<SingleAlternativeQuery> q,
  const std::map<std::uint64_t, size_t>& nodeIdToPos)
{
  // get left and right index
  if (join["left"].isUInt64() && join["right"].isUInt64())
  {
    auto leftID = join["left"].asUInt64();
    auto rightID = join["right"].asUInt64();

    auto itLeft = nodeIdToPos.find(leftID);
    auto itRight = nodeIdToPos.find(rightID);

    if (itLeft != nodeIdToPos.end() && itRight != nodeIdToPos.end())
    {

      auto op = join["op"].asString();
      if (op == "Precedence")
      {
        auto minDist = join["minDistance"].asUInt();
        auto maxDist = join["maxDistance"].asUInt();
        if(join["segmentation-name"].isString() && join["segmentation-name"].asString() != "")
        {
          q->addOperator(std::make_shared<Precedence>(db, getGraphStorageFunc, join["segmentation-name"].asString(),
            minDist, maxDist),
            itLeft->second, itRight->second);
        }
        else
        {
          q->addOperator(std::make_shared<Precedence>(db, getGraphStorageFunc,
            minDist, maxDist),
            itLeft->second, itRight->second);
        }
      }
      else if (op == "Inclusion")
      {
        q->addOperator(std::make_shared<Inclusion>(db, getGraphStorageFunc), itLeft->second, itRight->second);
      }
      else if (op == "Overlap")
      {
        q->addOperator(std::make_shared<Overlap>(db, getGraphStorageFunc), itLeft->second, itRight->second);
      }
      else if (op == "IdenticalCoverage")
      {
        q->addOperator(std::make_shared<IdenticalCoverage>(db, getGraphStorageFunc), itLeft->second, itRight->second);
      }
      else if (op == "Dominance")
      {

        std::string name = join["name"].isString() ? join["name"].asString() : "";

        if (join["edgeAnnotations"].isArray() && join["edgeAnnotations"].size() > 0)
        {
          auto anno = getEdgeAnno(db, join["edgeAnnotations"][0]);
          q->addOperator(std::make_shared<Dominance>(name, getAllGraphStorageFunc, db.strings, anno),
            itLeft->second, itRight->second);

        }
        else
        {

          auto minDist = join["minDistance"].asUInt();
          auto maxDist = join["maxDistance"].asUInt();
          if (minDist == 0 && maxDist == 0)
          {
            // unlimited range
            minDist = 1;
            maxDist = uintmax;
          }

          q->addOperator(std::make_shared<Dominance>(name, getAllGraphStorageFunc, db.strings,
            minDist, maxDist),
            itLeft->second, itRight->second);
        }
      }
      else if (op == "Pointing")
      {

        std::string name = join["name"].asString();

        if (join["edgeAnnotations"].isArray() && join["edgeAnnotations"].size() > 0)
        {
          auto anno = getEdgeAnno(db, join["edgeAnnotations"][0]);
          q->addOperator(std::make_shared<Pointing>(name, getAllGraphStorageFunc, db.strings, anno),
            itLeft->second, itRight->second);

        }
        else
        {

          auto minDist = join["minDistance"].asUInt();
          auto maxDist = join["maxDistance"].asUInt();

          if (minDist == 0 && maxDist == 0)
          {
            // unlimited range
            minDist = 1;
            maxDist = uintmax;
          }

          q->addOperator(std::make_shared<Pointing>(name, getAllGraphStorageFunc, db.strings, minDist, maxDist),
            itLeft->second, itRight->second);
        }
      }

    }

  }
}

Annotation JSONQueryParser::getEdgeAnno(const DB& db, const Json::Value& edgeAnno)
{

  std::uint32_t ns = 0;
  std::uint32_t name = 0;
  std::uint32_t value = 0;

  if (edgeAnno["textMatching"].asString() == "EXACT_EQUAL")
  {
    if (edgeAnno["namespace"].isString())
    {
      std::string nsStr = edgeAnno["namespace"].asString();
      auto search = db.strings.findID(nsStr);
      // if string is not found set to an invalid value
      ns = search.first ? search.second : std::numeric_limits<std::uint32_t>::max();
    }
    if (edgeAnno["name"].isString())
    {
      std::string nameStr = edgeAnno["name"].asString();
      auto search = db.strings.findID(nameStr);
      // if string is not found set to an invalid value
      name = search.first ? search.second : std::numeric_limits<std::uint32_t>::max();
    }
    if (edgeAnno["value"].isString())
    {
      std::string valueStr = edgeAnno["value"].asString();
      auto search = db.strings.findID(valueStr);
      // if string is not found set to an invalid value
      value = search.first ? search.second : std::numeric_limits<std::uint32_t>::max();
    }
  }
  // TODO: what about regex?

  return Init::initAnnotation(name, value, ns);
}

bool JSONQueryParser::canReplaceRegex(const std::string& str) 
{
  // Characters that have a meaning according to
  // https://github.com/google/re2/wiki/Syntax
  // Characters used in not supported functions are not included.
  if(str.find_first_of(".[]\\|*+?{}()^$") == std::string::npos)
  {
    // No meta character found in string, might be replaced    
    RE2 regex(str);
    if(regex.ok())
    {
      return true;
    }
    else
    {
      // If there is an error during parsing this is still a regex (an invalid one).
      // Treating it like a exact string would not give the same result.
      return false;
    }
  }
  else
  {
    // contains special regex characters
    return false;
  }
}


JSONQueryParser::~JSONQueryParser()
{
}

File: src/parser/jsonqueryparser.rs
use serde_json;

use exec::nodesearch::NodeSearchSpec;
use graphdb::GraphDB;
use operator::edge_op::EdgeAnnoSearchSpec;
use query::conjunction::Conjunction;
use query::disjunction::Disjunction;

use operator::{
    DominanceSpec, IdenticalCoverageSpec, IdenticalNodeSpec, InclusionSpec, OperatorSpec,
    OverlapSpec, PartOfSubCorpusSpec, PointingSpec, PrecedenceSpec,
};

use std::collections::BTreeMap;

pub fn parse<'a>(query_as_string: &str, db: &GraphDB) -> Option<Disjunction<'a>> {
    let root: serde_json::Value = serde_json::from_str(query_as_string).ok()?;

    let mut conjunctions: Vec<Conjunction> = Vec::new();
    // iterate over all alternatives
    let alternatives = root.get("alternatives")?.as_array()?;

    for alt in alternatives.iter() {
        let mut q = Conjunction::new();

        let mut first_node_pos: Option<usize> = None;

        // add all nodes
        let mut node_id_to_pos: BTreeMap<usize, usize> = BTreeMap::new();
        if let &serde_json::Value::Object(ref nodes) = alt.get("nodes")? {
            for (node_name, node) in nodes.iter() {
                if let Some(node_obj) = node.as_object() {
                    if let Ok(ref node_id) = node_name.parse::<u64>() {
                        let node_id = node_id.clone() as usize;

                        let pos = parse_node(node_obj, &mut q)?;
                        if first_node_pos.is_none() {
                            first_node_pos = Some(pos.clone());
                        }
                        node_id_to_pos.insert(node_id, pos);
                    }
                }
            }
        }

        // add all joins
        if let &serde_json::Value::Array(ref joins) = alt.get("joins")? {
            for j in joins.iter() {
                if let &serde_json::Value::Object(ref j_obj) = j {
                    parse_join(j_obj, &mut q, &node_id_to_pos, db);
                }
            }
        }

        // add all meta-data
        if let Some(meta_obj) = alt.get("meta") {
            let mut first_meta_idx: Option<usize> = None;

            if let Some(meta_array) = meta_obj.as_array() {
                for m in meta_array.iter() {
                    // add an artificial node that describes the document/corpus node
                    if let Some(meta_node_idx) = add_node_annotation(
                        &mut q,
                        m.get("namespace").and_then(|n| n.as_str()),
                        m.get("name").and_then(|n| n.as_str()),
                        m.get("value").and_then(|n| n.as_str()),
                        is_regex(m),
                        true,
                        None,
                    ) {
                        if let Some(first_meta_idx) = first_meta_idx {
                            // avoid nested loops by joining additional meta nodes with a "identical node"
                            q.add_operator(
                                Box::new(IdenticalNodeSpec {}),
                                first_meta_idx,
                                meta_node_idx,
                            );
                        } else if let Some(first_node_pos) = first_node_pos {
                            first_meta_idx = Some(meta_node_idx);
                            // add a special join to the first node of the query
                            q.add_operator(
                                Box::new(PartOfSubCorpusSpec::new(1, usize::max_value())),
                                first_node_pos,
                                meta_node_idx,
                            );
                            // Also make sure the matched node is actually a document
                            // (the @* could match anything in the hierarchy, including the toplevel corpus)
                            if let Some(doc_anno_idx) = add_node_annotation(
                                &mut q,
                                Some("annis"),
                                Some("doc"),
                                None,
                                false,
                                true,
                                None,
                            ) {
                                q.add_operator(
                                    Box::new(IdenticalNodeSpec {}),
                                    meta_node_idx,
                                    doc_anno_idx,
                                );
                            }
                        }
                    }
                }
            }
        }

        conjunctions.push(q);
    }

    if !conjunctions.is_empty() {
        return Some(Disjunction::new(conjunctions));
    }

    return None;
}

fn parse_node(
    node: &serde_json::Map<String, serde_json::Value>,
    q: &mut Conjunction,
) -> Option<usize> {
    let variable = node.get("variable").and_then(|s| s.as_str());
    // annotation search?
    if node.contains_key("nodeAnnotations") {
        if let serde_json::Value::Array(ref a) = node["nodeAnnotations"] {
            if !a.is_empty() {
                // get the first one
                let a = &a[0];
                return add_node_annotation(
                    q,
                    a.get("namespace").and_then(|n| n.as_str()),
                    a.get("name").and_then(|n| n.as_str()),
                    a.get("value").and_then(|n| n.as_str()),
                    is_regex(a),
                    false,
                    variable,
                );
            }
        }
    }

    // check for special non-annotation search constructs
    // token search?
    if (node.contains_key("spannedText") && node["spannedText"].is_string())
        || (node.contains_key("token")
            && node["token"].is_boolean()
            && node["token"].as_bool() == Some(true))
    {
        let spanned = node.get("spannedText").and_then(|s| s.as_str());

        if let Some(tok_val) = spanned {
            let mut leafs_only = false;
            if let Some(is_token) = node["token"].as_bool() {
                if is_token {
                    // special treatment for explicit searches for token (tok="...)
                    leafs_only = true;
                }
            }
            if node.contains_key("spanTextMatching")
                && node["spanTextMatching"].as_str() == Some("REGEXP_EQUAL")
            {
                return Some(q.add_node(
                    NodeSearchSpec::RegexTokenValue {
                        val: String::from(tok_val),
                        leafs_only,
                    },
                    variable,
                ));
            } else {
                return Some(q.add_node(
                    NodeSearchSpec::ExactTokenValue {
                        val: String::from(tok_val),
                        leafs_only,
                    },
                    variable,
                ));
            }
        } else {
            return Some(q.add_node(NodeSearchSpec::AnyToken, variable));
        }
    } else {
        // just search for any node
        return Some(q.add_node(NodeSearchSpec::AnyNode, variable));
    }
}

fn parse_join(
    join: &serde_json::Map<String, serde_json::Value>,
    q: &mut Conjunction,
    node_id_to_pos: &BTreeMap<usize, usize>,
    db: &GraphDB,
) {
    // get left and right index
    if let (Some(left_id), Some(right_id)) = (
        join.get("left").and_then(|n| n.as_u64()),
        join.get("right").and_then(|n| n.as_u64()),
    ) {
        let left_id = left_id as usize;
        let right_id = right_id as usize;
        if let (Some(pos_left), Some(pos_right)) =
            (node_id_to_pos.get(&left_id), node_id_to_pos.get(&right_id))
        {
            let spec_opt: Option<Box<OperatorSpec>> = match join.get("op").and_then(|s| s.as_str())
            {
                Some("Precedence") => {
                    let min_dist = join.get("minDistance").and_then(|n| n.as_u64());
                    let max_dist = join.get("maxDistance").and_then(|n| n.as_u64());
                    let seg_name = join.get("segmentation-name").and_then(|s| s.as_str());

                    let spec = PrecedenceSpec {
                        segmentation: seg_name.map(|s| String::from(s)),
                        min_dist: min_dist.unwrap_or(1) as usize,
                        max_dist: max_dist.unwrap_or(1) as usize,
                    };
                    Some(Box::new(spec))
                }
                Some("IdenticalCoverage") => {
                    let spec = IdenticalCoverageSpec {};
                    Some(Box::new(spec))
                }
                Some("Inclusion") => {
                    let spec = InclusionSpec {};
                    Some(Box::new(spec))
                }
                Some("Overlap") => {
                    let spec = OverlapSpec {};
                    Some(Box::new(spec))
                }
                Some("Dominance") => {
                    let min_dist = join
                        .get("minDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;
                    let max_dist = join
                        .get("maxDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;

                    let (min_dist, max_dist) = if min_dist == 0 && max_dist == 0 {
                        // unlimited range
                        (1, usize::max_value())
                    } else {
                        (min_dist, max_dist)
                    };

                    let name = join.get("name").and_then(|n| n.as_str());
                    let edge_anno = join
                        .get("edgeAnnotations")
                        .and_then(|a| a.as_array())
                        .and_then(|a| get_edge_anno(&a[0]));
                    let spec =
                        DominanceSpec::new(db, name.unwrap_or(""), min_dist, max_dist, edge_anno);
                    Some(Box::new(spec))
                }
                Some("Pointing") => {
                    let min_dist = join
                        .get("minDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;
                    let max_dist = join
                        .get("maxDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;

                    let name = join.get("name").and_then(|n| n.as_str());
                    let edge_anno = join
                        .get("edgeAnnotations")
                        .and_then(|a| a.as_array())
                        .and_then(|a| get_edge_anno(&a[0]));

                    let (min_dist, max_dist) = if min_dist == 0 && max_dist == 0 {
                        // unlimited range
                        (1, usize::max_value())
                    } else {
                        (min_dist, max_dist)
                    };

                    let spec =
                        PointingSpec::new(db, name.unwrap_or(""), min_dist, max_dist, edge_anno);
                    Some(Box::new(spec))
                }
                Some("PartOfSubcorpus") => {
                    let min_dist = join
                        .get("minDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;
                    let max_dist = join
                        .get("maxDistance")
                        .and_then(|n| n.as_u64())
                        .unwrap_or(1) as usize;

                    let (min_dist, max_dist) = if min_dist == 0 && max_dist == 0 {
                        // unlimited range
                        (1, usize::max_value())
                    } else {
                        (min_dist, max_dist)
                    };

                    let spec = PartOfSubCorpusSpec::new(min_dist, max_dist);
                    Some(Box::new(spec))
                }
                Some("IdenticalNode") => Some(Box::new(IdenticalNodeSpec)),
                // TODO: add more operators
                _ => None,
            };
            if let Some(spec) = spec_opt {
                q.add_operator(spec, pos_left.clone() as usize, pos_right.clone() as usize);
            }
        }
    }
}

fn get_edge_anno(json_node: &serde_json::Value) -> Option<EdgeAnnoSearchSpec> {
    if let Some(tm) = json_node.get("textMatching").and_then(|n| n.as_str()) {
        if tm == "EXACT_EQUAL" {
            let name = json_node.get("name")?.as_str()?;

            return Some(EdgeAnnoSearchSpec::ExactValue {
                ns: json_node
                    .get("namespace")
                    .and_then(|n| n.as_str())
                    .map(|s| String::from(s)),
                val: json_node
                    .get("value")
                    .and_then(|n| n.as_str())
                    .map(|s| String::from(s)),
                name: String::from(name),
            });
        }
        // TODO: what about regex?
    }
    None
}

fn is_regex(json_node: &serde_json::Value) -> bool {
    if let Some(tm) = json_node.get("textMatching").and_then(|n| n.as_str()) {
        if tm == "REGEXP_EQUAL" {
            return true;
        }
    }
    return false;
}

fn add_node_annotation(
    q: &mut Conjunction,
    ns: Option<&str>,
    name: Option<&str>,
    value: Option<&str>,
    regex: bool,
    is_meta: bool,
    variable: Option<&str>,
) -> Option<usize> {
    if let Some(name_val) = name {
        // TODO: replace regex with normal text matching if this is not an actual regular expression

        // search for the value
        if regex {
            if let Some(val) = value {
                let mut n: NodeSearchSpec = NodeSearchSpec::new_regex(ns, name_val, val, is_meta);
                return Some(q.add_node(n, variable));
            }
        } else {
            // has namespace?
            let mut n: NodeSearchSpec = NodeSearchSpec::new_exact(ns, name_val, value, is_meta);
            return Some(q.add_node(n, variable));
        }
    }
    return None;
}
